net/mwan3/files/lib/mwan3/mwan3.sh

```diff
@@ -338,6 +338,16 @@ mwan3_set_general_iptables()
 			done
 		fi
 
+		if [ -n "${current##*-N mwan3_pre*}" ]; then
+			mwan3_push_update -N mwan3_pre
+			mwan3_push_update -A mwan3_pre \
+					  -m mark ! --mark "0x0/$MMX_MASK" \
+					  -j MARK --set-xmark "0x0/$MMX_MASK"
```

Comment:
mwan3_hook applies to both PREROUTING and OUTPUT. We need to reset the mark in PREROUTING only, that's why I created a new chain. If we add a rule that resets the mark to mwan3_hook at the first position, it will apply to OUTPUT and break mwan3 tracking and mwan3 use command.

Regarding that you don't seem to be affected, although you also have an 6in4 tunnel. The issue is that:
1) First outgoing packet gets interface mark (for example, 0x200) and is correctly routed **via mwan3**
2) Reply packet is already marked 0x3f00 before reaching mwan3_hook, and mwan3_hook saves this mark and the whole connection gets mark 0x3f00
3) Second and all subsequent outgoing packets gets mark 0x3f00 (restored from connection mark by mwan3_hook) and goes out **via default routing table** 
You may not even notice that if you have only one IPv6 connection. But the things break when you have multiple connections.
Example: I have wan6 (native ipv6) and wanb6 (6in4). In the default routing table, wan6 has metric 1 and wanb6 has metric 2. And I have a mwan3 rule that routes outgoing connections to the specific resource via wanb6.
First outgoing packet to this resource is correctly routed **via mwan3** and goes out from wanb6.
Second and all subsequent outgoing packets goes out **via default routing table** via the wrong interface (wan6), resulting in the packet being dropped.

Reply:
The same applies to ipv4 ipsec tunnels - second and all subsequent outgoing packets goes out via default routing table via the main interface instead of being routed by mwan3 via the ipsec interface.

Reply:
The test for `! --mark "0x0/$MMX_MASK"` is redundant. "If it is not zero, reset it to zero" and "reset it to zero no matter what" are 100% equivalent instructions.

Reply:
> The inner packet passes through the tables first, getting the right mark for the tunnel interface. Then it gets encapsulated, and the outer packet passes through the tables again, preserving the inner packet's mark. As the mark is already set, all iptables rules will be skipped. When it comes to making a routing decision, the right routing table won't be selected (because the mark points to the table for the tunnel interface, not for the WAN netdev). In the lucky case, it will be handled by the main table, in the unlucky case (ipv4-in-ipv4?) it may fail to be routed.

You're right. There is an issue with egress too. It works exactly as you described.
But I don't know what to do with that. If we clear the mark for outgoing packets, routing will work as expected, but it will break mwan3 tracking and 'mwan3 use' command which relies on the pre-set mark.

Reply:
Probably the best way would be to solve the root of the issue, i.e. to patch the linux kernel to remove retaining the packet mark after de/encapsulation. I think retaining the mark is ridiculous and not technically correct, because inner packet is not the same packet as the tunnel packet and vice versa, and they belong to the different connections, and this behavior breaks routing based on marks,
However, I'm afraid that I don't have enough knowledge to patch the linux kernel.
I believe this issue affects all kernel-based tunnels - 6in4, l2tp, ipsec, wireguard etc.

Reply:
Thanks for the suggestion, I've updated the PR. Made some tests and it seems that if we clear the mark in POSTROUTING, it doesn't propagate to the tunnel connection itself. 'mwan3 use' works fine too.